### PR TITLE
Usbmanager spec.sh

### PR DESCRIPTION
**Pull request by christoph-zededa**
* improve speed by minimizing the amount of execves in order to make developing faster
* add usb devices
* installer: generate usb model manifest
* add parent group to model manifest (currently disabled)
* ignore NICs but allow modems

### CODE REVIEW

**Review by eriknordmark — APPROVED** [review]
LGTM

**Review by eriknordmark — CHANGES_REQUESTED** [review]
I see lots of commits including fome rrverts. Does it make sense to squash them to one ir a few?

**Comment by christoph-zededa** [comment]
putting to WIP as I have to do some more testing

**Comment by christoph-zededa** [comment]
> I see lots of commits including fome rrverts. Does it make sense to squash them to one ir a few?

I understand that it makes it sometimes more time-intensive to read as it is more code to read, but recently I bisected several issues and both times it was a small commit (that would have nice being squashed) and it helped to revert only a small commit instead of a big one.

File: pkg/debug/spec.sh
Comment (eriknordmark) [inline pkg/debug/spec.sh]: This can cause significant truncation for devices which only have a few GB of memory, since Linux typically doesn't get all of the 2G/4G/8G with the BIOS stealing some of it.
Hence you'll truncate to 1G/3G/7G which is a significant loss.

Was there a reason to truncate to GB?

Comment (rene) [reply pkg/debug/spec.sh]: This is the reason to be in KB: https://github.com/lf-edge/eve/commit/7ef8cd37896b3f03685209edf70533ba15cc8dfa

@christoph-zededa , which errors are you referring to? I didn't experience any error while testing these changes.

Comment (eriknordmark) [reply pkg/debug/spec.sh]: @rene I agree the unit should be in MB, but appearently there is a bug in zcli which can't handle the trailing "M" but if you give it a number it is correctly interpreted as being in MB.

**Comment by eriknordmark** [comment]
If we want to work with the (currently broken) zcli then the trailing "M" needs to be dropped.

File: pkg/debug/spec.sh
Comment (eriknordmark) [reply pkg/debug/spec.sh]: GB causes too much loss of accuracy. Hence we should keep this in MB.
But need to drop the 'M' at the end of the memory number since zcli is currently broken if you add that.

**Review by eriknordmark — APPROVED** [review]
LGTM - let's worry about the MB vs GB unit separately since that appears to be painful.
Some conflicts to resolve though.

File: pkg/debug/spec.sh
Comment (milan-zededa) [inline pkg/debug/spec.sh:460]: Seems like an obsolete and no longer appropriate comment

File: pkg/debug/spec.sh
Comment (milan-zededa) [inline pkg/debug/spec.sh:369]: Shouldn't this be IO_TYPE_USB (for non-modem device) or IO_TYPE_WWAN (for cellular modem)?

File: pkg/debug/spec.sh
Comment (milan-zededa) [inline pkg/debug/spec.sh:383]: Without parameter `-u` cellular modems will not be listed (?)
IMHO they should be always included with IO_TYPE_WWAN.

File: pkg/debug/spec.sh
Comment (milan-zededa) [inline pkg/debug/spec.sh:378]: What about adding VendorID:ProductID as `description` when verbose is enabled? Although not recognized and not allowed by the controller, it can be used by the user to determine which device each entry represents. That was actually the reason why AT commands were used to get some info for each modem. With multiple modems, it can be difficult to tell them apart and address of modem attachment is not something that users can easily map to hardware. (modem IMEI would be ideal but it is difficult to get from a simple shell script)

File: pkg/mkimage-raw-efi/install
Comment (milan-zededa) [inline pkg/mkimage-raw-efi/install:495]: Does it make sense to continue supporting device models assigning USB devices by PCI controllers?

File: pkg/debug/spec.sh
Comment (milan-zededa) [inline pkg/debug/spec.sh:464]: What if `-u` is not passed to the script. Will USB NIC be skipped altogether?

**Review by milan-zededa — COMMENTED** [review]
Apologies for reviewing this after the merge. I was waiting for the WIP to disappear, but then it got merged shortly after.
I will test this tomorrow with my modems and report any discrepancies between the previous and the new version of spec.sh.

File: pkg/debug/spec.sh
Comment (christoph-zededa) [reply pkg/debug/spec.sh:369]: `IO_TYPE_USB` will be added in a later PR once parent assign groups are supported.

Regarding `IO_TYPE_WWAN` I am unsure; why would we care what kind of usb device we passthrough into the edge application?

Besides `IO_TYPE_WWAN` I also found `IO_TYPE_LTE`.

Comment (milan-zededa) [reply pkg/debug/spec.sh:369]: Note that listing modems by spec.sh is not only for passthrough purposes. Controller needs to now about all network adapters (IO_TYPE_ETH + IO_TYPE_WWAN), so that they can have network config attached and used for device connectivity purposes.
And ignore IO_TYPE_LTE, this was apparently superseded by IO_TYPE_WWAN (LTE is only one generation of cellular connectivity, but 3G or 5G can be used as well, so LTE is not a generic enough label for modems).

File: pkg/debug/spec.sh
Comment (christoph-zededa) [reply pkg/debug/spec.sh:378]: As I see it, `picocom` was used to get the description of the device, like product name.

Now the product name is set in the `phylabel`, e.g.:
```
    {
      "ztype": "IO_TYPE_WWAN",
      "phylabel": "SierraWirelessEM7565QualcommSnapdragonX16LTE-A",
      "assigngrp": "",
      "cost": 10,
      "phyaddrs": {
        "usbaddr": "3:1.4"
      },
      "logicallabel": "SierraWirelessEM7565QualcommSnapdragonX16LTE-A",
      "usagePolicy": {}
    },
```

https://github.com/lf-edge/eve/pull/3555/files#diff-5082c5ce36254a19dd88fff26f3d05bc815ca75f4efb1a8b0d4868803ac6da51R335 <- that's how it gets the product string
But still, if we have the same modem twice, the `phylabel` would be indistinguishable, so I will add the bus and port number, so it will look like: `SierraWirelessEM7565QualcommSnapdragonX16LTE-A-USB3-1.4`

Comment (milan-zededa) [reply pkg/debug/spec.sh:378]: Note that logicallabel is used by the controller as the adapter name and it will appear in UI. So for that we should generate something shorter for nicer visualization, like modem0, modem1, etc. I think phylabel can be long like that (but we should check with the controller team if there are any string length restrictions imposed in the database).

Comment (milan-zededa) [reply pkg/debug/spec.sh:378]: But I also think that using raw and shorter "VendorID:ProductID" (+ address) for physical label would be fine as well.

Comment (christoph-zededa) [reply pkg/debug/spec.sh:378]: I though `phylabel` and `logicallabel` have to be the same string - I ran into issues with them being different.

I would actually prefer seeing the name of the modem over just `modem0`.

Comment (milan-zededa) [reply pkg/debug/spec.sh:378]: It is expected that the user will edit the generated device model anyway and put there some label relevant to their deployment. Logical label should really be something short that can be displayed in UI without taking half the screen.
Physical label can be different, at least in newer EVE versions it is not a problem anymore.

File: pkg/debug/spec.sh
Comment (christoph-zededa) [reply pkg/debug/spec.sh:460]: What about devices that are not connected via usb?

Comment (milan-zededa) [reply pkg/debug/spec.sh:460]: We do not support cellular modems connected over PCI yet.
As for WiFi, `if [ -d "$ETH/wireless" ]` used below should work regardless of the bus type.

**Pull request merged**
